ml-prims/src/linalg/qr.h

```diff
@@ -31,39 +32,38 @@ namespace LinAlg {
  * @param n_rows: number rows of input matrix
  * @param n_cols: number columns of input matrix
  * @param cusolverH cusolver handle
- * @param mgr device allocator for temporary buffers during computation
+ * @param stream cuda stream
+ * @param allocate device allocator for temporary buffers during computation
  * @{
  */
 template <typename math_t>
-void qrGetQ(math_t *&M, math_t *&Q, int n_rows, int n_cols,
+void qrGetQ(math_t *M, math_t *Q, int n_rows, int n_cols,
             cusolverDnHandle_t cusolverH, cudaStream_t stream,
-            DeviceAllocator &mgr) {
+            std::shared_ptr<deviceAllocator> allocator) {
   int m = n_rows, n = n_cols;
   int k = min(m, n);
   CUDA_CHECK(cudaMemcpyAsync(Q, M, sizeof(math_t) * m * n,
                              cudaMemcpyDeviceToDevice, stream));
 
-  math_t *tau = (math_t *)mgr.alloc(sizeof(math_t) * k);
-  CUDA_CHECK(cudaMemsetAsync(tau, 0, sizeof(math_t) * k, stream));
+  device_buffer<math_t> tau(allocator, stream, k);
+  CUDA_CHECK(cudaMemsetAsync(tau.data(), 0, sizeof(math_t) * k, stream));
 
-  int *devInfo = (int *)mgr.alloc(sizeof(int));
+  device_buffer<int> devInfo(allocator, stream, 1);
   int Lwork;
 
   CUSOLVER_CHECK(cusolverDngeqrf_bufferSize(cusolverH, m, n, Q, m, &Lwork));
-  math_t *workspace = (math_t *)mgr.alloc(sizeof(math_t) * Lwork);
+  device_buffer<math_t> workspace(allocator, stream, Lwork);
   CUSOLVER_CHECK(
-    cusolverDngeqrf(cusolverH, m, n, Q, m, tau, workspace, Lwork, devInfo, stream));
-  mgr.free(workspace, stream);
+      cusolverDngeqrf(cusolverH, m, n, Q, m, tau.data(), workspace.data(), Lwork, devInfo.data(), stream));
+  ///@todo: without deviceSynchronize *SquareMatrixNorm* ml-prims unit-tests fail
+  /// I've not been able to track down the rabbit hole of function calls to root-cause the
+  /// reason. Hence, as a temporary fix, leaving a device-sync here! :(
+  CUDA_CHECK(cudaDeviceSynchronize());
   CUSOLVER_CHECK(
-    cusolverDnorgqr_bufferSize(cusolverH, m, n, k, Q, m, tau, &Lwork));
-  workspace = (math_t *)mgr.alloc(sizeof(math_t) * Lwork);
+      cusolverDnorgqr_bufferSize(cusolverH, m, n, k, Q, m, tau.data(), &Lwork));
+  device_buffer<math_t> workspace1(allocator, stream, Lwork);

```

Comment:
these 2 are dependent calls and thus safer to compute their workspace sizes separately. This separate workspace was a remnant of my debugging related to the below mentioend cudaDeviceSynchronize thingy. Fixed the workspace reuse.

Reply:
If the workspace can be reused as the calls are dependent I don't see why it is safer to first allocate with one size and then allocate later with a different size instead of just directly allocate with the larger of the two sizes. Mainly out of curiosity what is your concern here?

Reply:
that's what I did in my latest commit :)